### PR TITLE
support apollo configuration center for datasource 

### PR DESCRIPTION
**Pull request by liyue201**
### Describe what this PR does / why we need it
[Apollo ](https://github.com/ctripcorp/apollo) is a very common and useful microservice configuration center.
### Does this pull request fix one issue?
no

### Describe how you did it
Refer to  the rep [apollo](https://github.com/ctripcorp/apollo) and [agollo](github.com/philchia/agollo)

### Describe how to verify it
run go test

### CODE REVIEW

**Comment by gorexlv** [comment]
Thanks for your job. Please create an issue first if possible.

**Comment by gorexlv** [comment]
ref #62

**Comment by gorexlv** [comment]
😄 why close this issue?

**Comment by liyue201** [comment]
ci没过，改个bug，在重新提交

**Comment by gorexlv** [comment]
> ci没过，改个bug，在重新提交

可以直接在这个PR上提交变更，不用新建PR

File: pkg/datasource/apollo/apollo.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"sync"
+)
+
+type apolloDataSource struct {
+	client      *agollo.Client
+	namespace   string
+	propertyKey string
+	changed     chan struct{}
+	quit        chan struct{}
+	sync.Once
```
Comment (gorexlv) [inline pkg/datasource/apollo/apollo.go:16]: Would once sync.Once make better sense?

File: pkg/datasource/apollo/apollo_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/douyu/jupiter/pkg/datasource/apollo/mockserver"
+	"github.com/philchia/agollo"
+	"github.com/stretchr/testify/assert"
+	"log"
+	"sync"
+	"testing"
+	"time"
+)
+
+func setup() {
```
Comment (gorexlv) [inline pkg/datasource/apollo/apollo_test.go:13]: Would TestMain be better?

File: pkg/datasource/apollo/apollo.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+// ReadConfig read config content from apollo
+func (ap *apolloDataSource) ReadConfig() ([]byte, error) {
+	ap.Once.Do(func() {
+		ap.client.Start()
+		changedEvent := ap.client.WatchUpdate()
+		xgo.Go(func() {
+			ap.watch(changedEvent)
+		})
+	})
+	value := ap.client.GetStringValueWithNameSpace(ap.namespace, ap.propertyKey, "")
```
Comment (gorexlv) [inline pkg/datasource/apollo/apollo.go:41]: Is there a timeout control mechanism where getting value from apollo?

File: pkg/datasource/apollo/apollo.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		propertyKey: key,
+		changed:     make(chan struct{}),
+		quit:        make(chan struct{}),
+	}
+	return ap
+}
+
+// ReadConfig read config content from apollo
+func (ap *apolloDataSource) ReadConfig() ([]byte, error) {
+	ap.Once.Do(func() {
```
Comment (gorexlv) [inline pkg/datasource/apollo/apollo.go:34]: How about move initialize logic to NewDataSource?

**Comment by gorexlv** [comment]
Thanks for your job.
I have two questions, and I would be very grateful if you could help me:

- How to deal with apollo breakdown or block?
- How to deal with apollo restart? is there any mechanism to avoid repeatedly pulling the same configuration, such as data reversion?

File: pkg/datasource/apollo/apollo.go (hunk 1 above)
Comment (liyue201) [reply pkg/datasource/apollo/apollo.go:16]: yes

File: pkg/datasource/apollo/apollo.go (hunk 4 above)
Comment (liyue201) [reply pkg/datasource/apollo/apollo.go:34]: 可以

File: pkg/datasource/apollo/apollo.go (hunk 3 above)
Comment (liyue201) [reply pkg/datasource/apollo/apollo.go:41]: 这个库内部定时每两秒钟从applo获取新的配置，更新它的缓存。这函数是从缓存里面读取配置。 所以这个函数不需要超时机制

**Comment by liyue201** [comment]
> Thanks for your job.
> I have two questions, and I would be very grateful if you could help me:
> 
> * How to deal with apollo breakdown or block?
> * How to deal with apollo restart? is there any mechanism to avoid repeatedly pulling the same configuration where restart apollo, such as data reversion?

1 .  因为依赖的这个客户端库内部有缓存，所以applo重启或连接不到，还是能读到缓存的数据。
2.   这个的客户端的ChangeEvent触发的条件是缓存数据和新获取的数据不一致时。所以不会有重复的数据

**Comment by gorexlv** [comment]
Thanks.

更新以后，请点击 Resolve Conversation.

**Pull request merged**